File: sort.py
import sys
def Insert_sort(A) :
    
    for j, val in enumerate(A) :
        
        i = j - 1
        while i >= 0 and A[i] > A[i+1] :
            A[i],A[i+1] = A[i+1],A[i]
            i = i - 1
        
    # print(A)
    return A
def Select_sort(A) :
    for i in range(len(A)):
        minVal = A[i]
        minValIndex = i
        j = i + 1
        while j < len(A) :
            if A[j] < minVal :
                minVal = A[j]
                minValIndex = j
            j = j + 1
        A[i],A[minValIndex] = A[minValIndex],A[i]
    print(A)
def Merge_sort(A) :
    if len(A) < 2 :
        return A
    divNum = int(len(A)/2)
    X = A[0:divNum]
    Y = A[divNum:]
    if len(X) == 1 :
        gredientA = X
    else:
        gredientA = Merge_sort(X)
    if len(Y) == 1 :
        gredientB = Y
    else:
        gredientB = Merge_sort(Y)
    result_slice = []
    length_of_all = len(gredientA) + len(gredientB)
    gredientA.append(sys.maxsize)
    gredientB.append(sys.maxsize)
    pointA,pointB = 0,0
    for i in range(length_of_all):
        if gredientA[pointA] < gredientB[pointB] :
            result_slice.append(gredientA[pointA])
            pointA = pointA + 1
        else :
            result_slice.append(gredientB[pointB])
            pointB = pointB + 1
    return result_slice
if __name__ == '__main__' :
    a = [31, 41, 59, 26, 41, 58, 90, 100, 29, 11, 25]
    print(a)
    Insert_sort(a)
    Select_sort(a)
    print(Merge_sort(a))

File: sort.go
package main

import (
	"fmt"
	"math"
)

func main() {
	input := []int{58, 43, 26, 11, 18, 55, 34, 90, 100, 12, 500, 10000, 78, 29, 103, 1, 32, 8}
	B := Insert_sort(input)
	C := Selection_sort(input)
	fmt.Println(B)
	fmt.Println(C)
	fmt.Println("into merge_sort!")
	fmt.Println(Merge_sort(input))
}
func Insert_sort(A []int) []int {

	for j := 1; j < len(A); j++ {
		i := j - 1
		for i >= 0 && A[i] > A[i+1] {
			A[i], A[i+1] = A[i+1], A[i]
			i = i - 1
		}
	}
	return A
}
func Selection_sort(A []int) []int {
	minVal := 0
	for j := 0; j < len(A); j++ {
		minVal = A[j]
		minValIndex := j
		i := j + 1
		for i < len(A) {
			if A[i] < minVal {
				minVal = A[i]
				minValIndex = i

			}
			i = i + 1
		}
		if minValIndex != j {
			A[j], A[minValIndex] = A[minValIndex], A[j]
		}
	}
	return A
}
func Merge_sort(A []int) []int {
	//	fmt.Println("Into mergo_sort!")
	if len(A) < 2 {
		return A
	}
	//	fmt.Println(A)
	X := A[0 : len(A)/2]
	Y := A[len(A)/2:]
	n := []int{}
	m := []int{}
	//	fmt.Println(X, Y)
	resultSlice := []int{}
	switch len(X) {
	case 1:
		n = X
	default:
		n = Merge_sort(X)
	}
	switch len(Y) {
	case 1:
		m = Y
	default:
		m = Merge_sort(Y)
	}
	i, j := 0, 0
	tempN := n[i]
	tempM := m[j]
	for k := 0; k < len(n)+len(m); k++ {

		if tempN < tempM {
			resultSlice = append(resultSlice, n[i])

			i = i + 1
			if i < len(n) {
				tempN = n[i]
			} else {
				tempN = math.MaxInt64
			}
		} else {
			resultSlice = append(resultSlice, m[j])
			j = j + 1
			if j < len(m) {
				tempM = m[j]
			} else {
				tempM = math.MaxInt64
			}
		}
	}
	return resultSlice
	//	fmt.Println(X)
	//	fmt.Println(Y)
	//	return A
}
